### Carga de datos, etc

In [2]:
import os
import re

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import cv2
import PIL.Image as Image

from skimage import io, color, transform
from skimage.feature import hog

import time

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score, classification_report

In [3]:
gray_matrix = np.load('DatosDF\gray_matrix.npy', allow_pickle=True)
hog = np.load('DatosDF\hog.npy', allow_pickle=True)
estados = np.load('DatosDF\estados.npy', allow_pickle=True)
label = np.load('DatosDF\labels.npy', allow_pickle=True)

<>:1: SyntaxWarning: invalid escape sequence '\g'
<>:2: SyntaxWarning: invalid escape sequence '\h'
<>:3: SyntaxWarning: invalid escape sequence '\e'
<>:4: SyntaxWarning: invalid escape sequence '\l'
<>:1: SyntaxWarning: invalid escape sequence '\g'
<>:2: SyntaxWarning: invalid escape sequence '\h'
<>:3: SyntaxWarning: invalid escape sequence '\e'
<>:4: SyntaxWarning: invalid escape sequence '\l'
C:\Users\gonza\AppData\Local\Temp\ipykernel_7336\488994593.py:1: SyntaxWarning: invalid escape sequence '\g'
  gray_matrix = np.load('DatosDF\gray_matrix.npy', allow_pickle=True)
C:\Users\gonza\AppData\Local\Temp\ipykernel_7336\488994593.py:2: SyntaxWarning: invalid escape sequence '\h'
  hog = np.load('DatosDF\hog.npy', allow_pickle=True)
C:\Users\gonza\AppData\Local\Temp\ipykernel_7336\488994593.py:3: SyntaxWarning: invalid escape sequence '\e'
  estados = np.load('DatosDF\estados.npy', allow_pickle=True)
C:\Users\gonza\AppData\Local\Temp\ipykernel_7336\488994593.py:4: SyntaxWarning: invalid

In [4]:
df = pd.DataFrame({'gray_matrix': gray_matrix,'hog': hog,'estados': estados,'label': label})

In [5]:
dfxusar = df[df['estados'] == 1]
elresto = df[df['estados'] == 0]
dfxusar

gray_matrix  \
16     [[175, 166, 185, 198, 202, 213, 212, 209, 210,...   
23     [[26, 29, 30, 35, 45, 60, 108, 145, 161, 161, ...   
30     [[17, 18, 19, 19, 17, 15, 16, 17, 18, 20, 18, ...   
31     [[0, 0, 0, 0, 0, 0, 0, 0, 2, 3, 0, 1, 13, 37, ...   
45     [[254, 254, 254, 254, 254, 254, 254, 254, 254,...   
...                                                  ...   
35863  [[9, 38, 69, 99, 86, 102, 111, 105, 65, 58, 64...   
35869  [[253, 253, 254, 253, 170, 101, 105, 102, 85, ...   
35875  [[252, 247, 146, 93, 94, 89, 99, 94, 93, 83, 9...   
35878  [[33, 44, 49, 40, 35, 28, 23, 14, 9, 16, 17, 5...   
35884  [[124, 129, 125, 123, 129, 128, 127, 128, 131,...   

                                                     hog  estados     label  
16     [0.20444939, 0.034390625, 0.026585897, 0.01487...        1     angry  
23     [0.46974203, 0.15378156, 0.0056655, 0.0, 0.002...        1     angry  
30     [0.16117984, 0.10060929, 0.12280247, 0.0026138...        1     angry  
31     [0.17848918, 0.014774916, 0.0, 0.0, 0.0, 0.0, ...        1     angry  
45     [0.025827257, 0.01296422, 0.020755643, 0.0, 0....        1     angry  
...                                                  ...      ...       ...  
35863  [0.23134665, 0.3235038, 0.13396424, 0.05265233...        1  surprise  
35869  [0.4190711, 0.32569548, 0.0045682807, 0.002312...        1  surprise  
35875  [0.32172725, 0.29283354, 0.09000543, 0.0720573...        1  surprise  
35878  [0.20763314, 0.31089163, 0.1581842, 0.02511550...        1  surprise  
35884  [0.20560372, 0.04748373, 0.03934206, 0.0998374...        1  surprise  

[4611 rows x 4 columns]

In [6]:
X = np.array(dfxusar['hog'].values.tolist())
y = dfxusar['label']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=19011)

## RandomForest

In [ ]:
%%time
# Crear el modelo
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)

# Entrenar el modelo
rf_classifier.fit(X_train, y_train)

In [ ]:
%%time
y_pred_e= rf_classifier.predict(X_train)
accuracy = accuracy_score(y_train, y_pred_e)
print(f'Precisión del modelo: {accuracy}')

print(classification_report(y_train, y_pred_e))

In [ ]:
%%time
# Realizar predicciones en el conjunto de prueba
y_pred = rf_classifier.predict(X_test)

# Evaluar el modelo
accuracy = accuracy_score(y_test, y_pred)
print(f'Precisión del modelo: {accuracy}')

print(classification_report(y_test, y_pred))

Precisión del modelo: 0.8039003250270856
              precision    recall  f1-score   support

       angry       0.84      0.68      0.75       101
     disgust       0.81      0.57      0.67       112
        fear       0.75      0.71      0.73       156
       happy       0.83      0.99      0.90       149
     neutral       0.79      0.89      0.84       153
         sad       0.73      0.80      0.76       119
    surprise       0.90      0.90      0.90       133

    accuracy                           0.80       923
   macro avg       0.81      0.79      0.79       923\
weighted avg       0.80      0.80      0.80       923\

CPU times: total: 46.9 ms\
Wall time: 55.9 ms


## [Conjunto Pequeño] GridSearch - Optimizacion de parametros

### Primero

In [ ]:
%%time

# Define el modelo de Random Forest
rf_classifier = RandomForestClassifier()

# Define una cuadrícula de hiperparámetros que deseas explorar
param_grid = {
    'n_estimators': [150,200,250,300],        # Número de árboles en el bosque
    'max_depth': [10,20,30],      # Profundidad máxima de los árboles
    'min_samples_split': [10,20,30],     # Mínimo de muestras requeridas para dividir un nodo
    'min_samples_leaf': [10,20,30]       # Mínimo de muestras requeridas en una hoja
}

# Realiza la búsqueda en cuadrícula con validación cruzada
grid_search = GridSearchCV(estimator=rf_classifier, param_grid=param_grid, cv=5, scoring='accuracy', verbose=10,n_jobs=8)
grid_search.fit(X_train, y_train)

# Muestra los mejores hiperparámetros encontrados
best_params = grid_search.best_params_
print(f"Mejores hiperparámetros: {best_params}")

# Entrena el modelo con los mejores hiperparámetros en todo el conjunto de entrenamiento
best_rf_classifier = grid_search.best_estimator_
best_rf_classifier.fit(X_train, y_train)

# Evalúa el modelo en el conjunto de prueba
y_pred = best_rf_classifier.predict(X_test)

# Evaluar el modelo
accuracy = accuracy_score(y_test, y_pred)
print(f'Precisión del modelo: {accuracy}')

print(classification_report(y_test, y_pred))

Fitting 5 folds for each of 108 candidates, totalling 540 fits\
Mejores hiperparámetros: {'max_depth': 30, 'min_samples_leaf': 10, 'min_samples_split': 20, 'n_estimators': 300}\
Precisión del modelo: 0.7963163596966414\
              precision    recall  f1-score   support

       angry       0.89      0.69      0.78       101
     disgust       0.83      0.45      0.58       112
        fear       0.77      0.71      0.74       156
       happy       0.79      1.00      0.88       149
     neutral       0.75      0.89      0.81       153
         sad       0.72      0.79      0.76       119
    surprise       0.89      0.95      0.92       133

    accuracy                           0.80       923
   macro avg       0.81      0.78      0.78       923\
weighted avg       0.80      0.80      0.79       923

CPU times: total: 2min 29s\
Wall time: 54min 49s

Resultados: 
- 'max_depth': 30
- 'min_samples_leaf': 10
- 'min_samples_split': 20
- 'n_estimators': 300


### Segundo

In [ ]:
#'max_depth': 30, 'min_samples_leaf': 10, 'min_samples_split': 20, 'n_estimators': 300

%%time
# Define el modelo de Random Forest
rf_classifier = RandomForestClassifier(n_estimators=300)

# Define una cuadrícula de hiperparámetros que deseas explorar
param_grid = {
    'max_depth': [25,30,45,60,75,90, None],      # Profundidad máxima de los árboles
    'min_samples_split': [14,16,18,20,22,24,26],     # Mínimo de muestras requeridas para dividir un nodo
    'min_samples_leaf': [4,6,8,10,12,14,16]       # Mínimo de muestras requeridas en una hoja
}

# Realiza la búsqueda en cuadrícula con validación cruzada
grid_search = GridSearchCV(estimator=rf_classifier, param_grid=param_grid, cv=5, scoring='accuracy', verbose=10,n_jobs=8)
grid_search.fit(X_train, y_train)

# Muestra los mejores hiperparámetros encontrados
best_params = grid_search.best_params_
print(f"Mejores hiperparámetros: {best_params}")

# Entrena el modelo con los mejores hiperparámetros en todo el conjunto de entrenamiento
best_rf_classifier = grid_search.best_estimator_
best_rf_classifier.fit(X_train, y_train)

# Evalúa el modelo en el conjunto de prueba
y_pred = best_rf_classifier.predict(X_test)

# Evaluar el modelo
accuracy = accuracy_score(y_test, y_pred)
print(f'Precisión del modelo: {accuracy}')

print(classification_report(y_test, y_pred))

Fitting 5 folds for each of 343 candidates, totalling 1715 fits
Mejores hiperparámetros: {'max_depth': None, 'min_samples_leaf': 4, 'min_samples_split': 14}
Precisión del modelo: 0.8049837486457205
              precision    recall  f1-score   support

       angry       0.85      0.70      0.77       101
     disgust       0.76      0.48      0.59       112
        fear       0.80      0.73      0.76       156
       happy       0.84      0.99      0.90       149
     neutral       0.76      0.90      0.83       153
         sad       0.72      0.80      0.76       119
    surprise       0.91      0.93      0.92       133

    accuracy                           0.80       923
   macro avg       0.80      0.79      0.79       923\
weighted avg       0.80      0.80      0.80       923\


**Resultados**
- 'max_depth': None
- 'min_samples_leaf': 4
- 'min_samples_split': 14

### Tercero

In [17]:
#'max_depth': 30, 'min_samples_leaf': 10, 'min_samples_split': 20, 'n_estimators': 300

# Define el modelo de Random Forest
rf_classifier = RandomForestClassifier(n_estimators=300)

# Define una cuadrícula de hiperparámetros que deseas explorar
param_grid = {
    'max_depth': [25,60, None],      
    'min_samples_split': [14,20,26],     
    'min_samples_leaf': [4,10,16],       
    'max_features' : [100,150,200],

    'criterion': ['gini','entropy']
}

# Realiza la búsqueda en cuadrícula con validación cruzada
grid_search = GridSearchCV(estimator=rf_classifier, param_grid=param_grid, cv=5, scoring='accuracy', verbose=10,n_jobs=12)
grid_search.fit(X_train, y_train)

# Muestra los mejores hiperparámetros encontrados
best_params = grid_search.best_params_
print(f"Mejores hiperparámetros: {best_params}")

# Entrena el modelo con los mejores hiperparámetros en todo el conjunto de entrenamiento
best_rf_classifier = grid_search.best_estimator_
best_rf_classifier.fit(X_train, y_train)

# Evalúa el modelo en el conjunto de prueba
y_pred = best_rf_classifier.predict(X_test)

# Evaluar el modelo
accuracy = accuracy_score(y_test, y_pred)
print(f'Precisión del modelo: {accuracy}')

print(classification_report(y_test, y_pred))

Fitting 5 folds for each of 162 candidates, totalling 810 fits
Mejores hiperparámetros: {'criterion': 'gini', 'max_depth': None, 'max_features': 200, 'min_samples_leaf': 4, 'min_samples_split': 14}
Precisión del modelo: 0.8136511375947996
              precision    recall  f1-score   support

       angry       0.85      0.70      0.77       101
     disgust       0.81      0.52      0.63       112
        fear       0.80      0.74      0.77       156
       happy       0.85      0.99      0.91       149
     neutral       0.78      0.90      0.84       153
         sad       0.70      0.82      0.75       119
    surprise       0.93      0.93      0.93       133

    accuracy                           0.81       923
   macro avg       0.82      0.80      0.80       923
weighted avg       0.82      0.81      0.81       923



**Resultados** 

Precisión del modelo: 0.8266522210184182
- 'criterion': 'gini', 
- 'max_depth': None, 
- 'max_features': 200, 
- 'min_samples_leaf': 4,
- 'min_samples_split': 14

### Cuarto

In [19]:
#'max_depth': 30, 'min_samples_leaf': 10, 'min_samples_split': 20, 'n_estimators': 300

# Define el modelo de Random Forest
rf_classifier = RandomForestClassifier(n_estimators=300)

# Define una cuadrícula de hiperparámetros que deseas explorar
param_grid = {
    'max_depth': [60, None],      
    'min_samples_split': [10,14],     
    'min_samples_leaf': [4,10],       
    'max_features' : [10,30,60],
    'n_estimators' : [100,300,600,900],
    'criterion': ['gini']
}

# Realiza la búsqueda en cuadrícula con validación cruzada
grid_search = GridSearchCV(estimator=rf_classifier, param_grid=param_grid, cv=5, scoring='accuracy', verbose=10,n_jobs=12)
grid_search.fit(X_train, y_train)

# Muestra los mejores hiperparámetros encontrados
best_params = grid_search.best_params_
print(f"Mejores hiperparámetros: {best_params}")

# Entrena el modelo con los mejores hiperparámetros en todo el conjunto de entrenamiento
best_rf_classifier = grid_search.best_estimator_
best_rf_classifier.fit(X_train, y_train)

# Evalúa el modelo en el conjunto de prueba
y_pred = best_rf_classifier.predict(X_test)

# Evaluar el modelo
accuracy = accuracy_score(y_test, y_pred)
print(f'Precisión del modelo: {accuracy}')

print(classification_report(y_test, y_pred))

Fitting 5 folds for each of 96 candidates, totalling 480 fits
Mejores hiperparámetros: {'criterion': 'gini', 'max_depth': None, 'max_features': 60, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 600}
Precisión del modelo: 0.8060671722643553
              precision    recall  f1-score   support

       angry       0.88      0.66      0.76       101
     disgust       0.83      0.49      0.62       112
        fear       0.78      0.74      0.76       156
       happy       0.82      1.00      0.90       149
     neutral       0.77      0.92      0.83       153
         sad       0.71      0.80      0.75       119
    surprise       0.90      0.92      0.91       133

    accuracy                           0.81       923
   macro avg       0.81      0.79      0.79       923
weighted avg       0.81      0.81      0.80       923



**Resultados** 

Precisión del modelo: 0.8060671722643553
- 'criterion': 'gini', 
- 'max_depth': None, 
- 'max_features': 60, 
- 'min_samples_leaf': 4,
- 'min_samples_split': 10
- 'n_estimators':600

### Quinto

In [20]:
rf_classifier = RandomForestClassifier(n_estimators=300)

param_grid = {
    'max_depth': [None],      
    'min_samples_split': [10,14],     
    'min_samples_leaf': [4],       
    'max_features' : [60,200],
    'n_estimators' : [300,600],
    'criterion': ['gini']
}

grid_search_thefifth = GridSearchCV(estimator=rf_classifier, param_grid=param_grid, cv=5, scoring='accuracy', verbose=10,n_jobs=12,return_train_score = True)
grid_search_thefifth.fit(X_train, y_train)

best_params = grid_search_thefifth.best_params_
print(f"Mejores hiperparámetros: {best_params}")

# Entrena el modelo con los mejores hiperparámetros en todo el conjunto de entrenamiento
best_rf_classifier = grid_search_thefifth.best_estimator_
best_rf_classifier.fit(X_train, y_train)

# Evalúa el modelo en el conjunto de prueba
y_pred = best_rf_classifier.predict(X_test)

# Evaluar el modelo
accuracy = accuracy_score(y_test, y_pred)
print(f'Precisión del modelo: {accuracy}')

print(classification_report(y_test, y_pred))

Fitting 5 folds for each of 8 candidates, totalling 40 fits
Mejores hiperparámetros: {'criterion': 'gini', 'max_depth': None, 'max_features': 200, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 600}
Precisión del modelo: 0.8114842903575298
              precision    recall  f1-score   support

       angry       0.84      0.67      0.75       101
     disgust       0.85      0.52      0.64       112
        fear       0.78      0.76      0.77       156
       happy       0.84      1.00      0.91       149
     neutral       0.78      0.90      0.84       153
         sad       0.70      0.82      0.75       119
    surprise       0.92      0.90      0.91       133

    accuracy                           0.81       923
   macro avg       0.82      0.80      0.80       923
weighted avg       0.82      0.81      0.81       923



**Resultados** 

Precisión del modelo: 0.8060671722643553
- 'criterion': 'gini', 
- 'max_depth':  None
- 'max_features':  200
- 'min_samples_leaf': 4
- 'min_samples_split': 10
- 'n_estimators': 600

In [21]:
results = grid_search_thefifth.cv_results_
acc_e =[]
acc_t=[]
n_nnn=[]
i=1
for mean_train_score, mean_test_score, params in zip(results['mean_train_score'], results['mean_test_score'], results['params']):
    acc_e.append(mean_train_score)
    acc_t.append(mean_test_score)
    n_nnn.append(i)
    print(f"Mean train score: {mean_train_score:.3f}, Mean test score: {mean_test_score:.3f} with parameters: {params}")
    i=i+2

Mean train score: 0.999, Mean test score: 0.795 with parameters: {'criterion': 'gini', 'max_depth': None, 'max_features': 60, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 300}
Mean train score: 0.999, Mean test score: 0.800 with parameters: {'criterion': 'gini', 'max_depth': None, 'max_features': 60, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 600}
Mean train score: 0.997, Mean test score: 0.793 with parameters: {'criterion': 'gini', 'max_depth': None, 'max_features': 60, 'min_samples_leaf': 4, 'min_samples_split': 14, 'n_estimators': 300}
Mean train score: 0.997, Mean test score: 0.795 with parameters: {'criterion': 'gini', 'max_depth': None, 'max_features': 60, 'min_samples_leaf': 4, 'min_samples_split': 14, 'n_estimators': 600}
Mean train score: 0.999, Mean test score: 0.796 with parameters: {'criterion': 'gini', 'max_depth': None, 'max_features': 200, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 300}
Mean train score: 0.998,

In [23]:
results = grid_search.cv_results_
acc_e =[]
acc_t=[]
n_nnn=[]
i=1
for  mean_test_score, params in zip( results['mean_test_score'], results['params']):
    acc_t.append(mean_test_score)
    n_nnn.append(i)
    print(f"Mean train score: {mean_train_score:.3f}, Mean test score: {mean_test_score:.3f} with parameters: {params}")
    i=i+2
    

Mean train score: 0.997, Mean test score: 0.761 with parameters: {'criterion': 'gini', 'max_depth': 60, 'max_features': 10, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 100}
Mean train score: 0.997, Mean test score: 0.787 with parameters: {'criterion': 'gini', 'max_depth': 60, 'max_features': 10, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 300}
Mean train score: 0.997, Mean test score: 0.791 with parameters: {'criterion': 'gini', 'max_depth': 60, 'max_features': 10, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 600}
Mean train score: 0.997, Mean test score: 0.791 with parameters: {'criterion': 'gini', 'max_depth': 60, 'max_features': 10, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 900}
Mean train score: 0.997, Mean test score: 0.767 with parameters: {'criterion': 'gini', 'max_depth': 60, 'max_features': 10, 'min_samples_leaf': 4, 'min_samples_split': 14, 'n_estimators': 100}
Mean train score: 0.997, Mean test 